tools/azsdk-cli/docs/specs/6-package-updater.spec.md

```diff
@@ -0,0 +1,689 @@
+
+Provides a singular CLI + MCP command that orchestrates the four tools (update-ci → update-changelog → update-version → update-metadata) and returns a combined JSON summary.
+
+### Detailed Design
+
+The four tools are intentionally small, composable units. Each emits a machine-readable JSON payload to enable deterministic chaining by an agent or CI workflow. All tools share common behaviors:
+
+| Area | Expected Behavior |
+|--------|-------------------|
+| Invocation Mode | CLI and MCP (agent) alias; same semantics |
+| Input Path Validation | Must exist and contain a recognizable SDK language structure; errors early if not |
+| Config Resolution | Load `swagger_to_sdk_config.json`; prefer `inline command` over `script path` when both are configured |
+| Timeout (default) | 5 minutes per tool invocation |
+| Output Schema | JSON with fields: result, message (see schema below) |
+| Idempotency | Multiple successive runs produce identical filesystem state except for timestamps/logs |
+| Script Implementation | MUST be PowerShell `.ps1`. SHOULD accept named parameters (e.g., `sdkRepoPath`). MUST use exit code `0` for success/partial/noop; non-zero only on outright failure. Legacy non-PS scripts should be wrapped by a small PowerShell shim. Scripts MUST write a concise stderr line prefixed with `ERROR:`; include suggested mitigation steps and links to docs in the message where possible. |
+| Script Execution Environment | All scripts are executed using PowerShell Core (`pwsh`) terminal |
+| Next Steps Hint | Plain language phrase embedded in `message` (e.g., `next_steps: update version`) |
+| Missing Required Tools | If a required external tool is not present (for example: `dotnet`, `gofmt`, `pwsh`), prompt the user to run the `verify-setup` tool which validates and documents missing prerequisites. |
+
+#### 1. Changelog Update Tool
+
+Name (CLI): `azsdk package update changelog`
+
+Name (MCP): `azsdk_package_update_changelog`
+
+Purpose: Ensure changelog is auto-generated and has an entry for the upcoming release (management-plane) or provide guidance (data-plane) where manual editing is expected.
+
+Inputs:
+
+- `--package-path <abs path>` (required)
+
+Execution Steps (Mgmt-Plane):
+
+1. Invoke changelog generation script or tool.
+2. Insert (or update) latest section with release date placeholder / actual date.
+3. Normalize section ordering / formatting.
+
+Execution Steps (Data-Plane):
+
+1. Return `noop` result advising manual edit and future version bump timing.
+
+Outputs:
+
+```json
+{"result":"succeeded", "message":"Changelog entry added for 2025-10-16", "next_steps": "update version"}
+```
+
+```json
+{"result":"noop", "message":"Data-plane changelog untouched; manual edits required", "next_steps": "update version"}
+```
+
+Failure Modes:
+
+- Missing `CHANGELOG.md` → create template (mgmt-plane) or noop (data-plane).
```

Comment:
I wonder if we should consider recovery here rather than continuing down the "yell at someone and make them fix it" when we have a problem.    Here, we could attempt to grab the latest changelog from the latest release tag and if we don't have a release, just create it.

Not saying that it needs to be this phase, but I think we need to shift our mindset about how we design tools.   They should be self-healing and find the pit of success when possible.   The approach of failing and making the caller's problem to fix is a big reason why our end-to-end experience has friction and why our team gets pulled into support.

//cc: @ronniegeraghty, @maririos for additional thoughts.

Reply:
Added the feedback to the spec.